### PR TITLE
[keepalived-vip] Add support to filter nodes in unicast and UDP services

### LINKED ISSUE #808: [keepalived-vip] unicast_peer should only include VRRP peers
With multi-node clusters, setting a label for restriction by nodeSelector in the daemonset spec and specifying `--use-unicast=true`, all service nodes are listed as unicast peers, rather than the subset of nodes that are specified by the nodeSelector.

```
 ✓ ( 11:27:05 ) ⦿ ryan@juniper
 ▶︎ ~/repos/vst/contrib/keepalived-vip ▶︎  master ● ▶︎ kubectl get no --show-labels
NAME            STATUS                     AGE       LABELS
10.10.34.71     Ready                      3d        kubernetes.io/hostname=10.10.34.71,type=vip
10.10.34.72     Ready                      3d        kubernetes.io/hostname=10.10.34.72,type=vip
10.10.34.73     Ready                      3d        kubernetes.io/hostname=10.10.34.73,type=vip
10.10.34.74     Ready                      3d        kubernetes.io/hostname=10.10.34.74
10.10.34.75     Ready                      3d        kubernetes.io/hostname=10.10.34.75
10.10.34.76     Ready                      3d        kubernetes.io/hostname=10.10.34.76
172.21.10.162   Ready,SchedulingDisabled   3d        kubernetes.io/hostname=172.21.10.162
 ✓ ( 11:27:12 ) ⦿ ryan@juniper
 ▶︎ ~/repos/vst/contrib/keepalived-vip ▶︎  master ● ▶︎ grep -A2 nodeSelector vip-daemonset.yaml
      nodeSelector:
        type: vip
 ✓ ( 11:27:29 ) ⦿ ryan@juniper
 ▶︎ ~/repos/vst/contrib/keepalived-vip ▶︎  master ● ▶︎ kubectl get pods
NAME                        READY     STATUS    RESTARTS   AGE       NODE
hostnames-1ni2l             1/1       Running   0          1h        10.10.34.74
hostnames-6rrss             1/1       Running   0          1h        10.10.34.73
hostnames-fmt0n             1/1       Running   0          1h        10.10.34.76
hostnames-lyb6w             1/1       Running   0          1h        10.10.34.71
hostnames-qo1xu             1/1       Running   0          1h        10.10.34.72
hostnames-zn7cn             1/1       Running   0          1h        10.10.34.75
kube-keepalived-vip-1hpkc   1/1       Running   0          53m       10.10.34.72
kube-keepalived-vip-75q8d   1/1       Running   0          53m       10.10.34.71
kube-keepalived-vip-t1egp   1/1       Running   0          53m       10.10.34.73
 ✓ ( 11:27:48 ) ⦿ ryan@juniper
 ▶︎ ~/repos/vst/contrib/keepalived-vip ▶︎  master ● ▶︎ kubectl exec kube-keepalived-vip-1hpkc cat /etc/keepalived/keepalived.conf


global_defs {
  vrrp_version 3
  vrrp_iptables KUBE-KEEPALIVED-VIP
}

vrrp_instance vips {
  state BACKUP
  interface bond0
  virtual_router_id 50
  priority 101
  nopreempt
  advert_int 1

  track_interface {
    bond0
  }


  unicast_src_ip 10.10.34.72
  unicast_peer {
    10.10.34.71
    10.10.34.73
    10.10.34.74
    10.10.34.75
    10.10.34.76
    172.21.10.162
  }
<SNIP>
}
```

### PR DESCRIPTION
**Pull request by aledbf**
closes #808

### CODE REVIEW

**Comment by aledbf** [comment]
ping @bprashanth

File: keepalived-vip/utils.go [hunk 1]
```diff
@@ -239,3 +271,81 @@ func parseNsName(input string) (string, string, error) {
...
+		return false, nil
+	}
+
+	return waitForPodCondition(kubeClient, ns, podName, condition, interval, timeout)
+}
+
+// waitForPodCondition waits for a pod in state defined by a condition (func)
+func waitForPodCondition(kubeClient *unversioned.Client, ns, podName string, condition func(pod *api.Pod) (bool, error),
+	interval, timeout time.Duration) error {
+	return wait.PollImmediate(interval, timeout, func() (bool, error) {
```
Comment (bprashanth) [inline keepalived-vip/utils.go]: this error is useless, suggest bubbling it up as something clearer. IT'll basically say: timed out waiting for condition, suggest something like: timed out waiting to observe own status as Running.

File: keepalived-vip/README.md [hunk 2]
```diff
@@ -24,19 +24,25 @@ With 2 or more instance of the pod running in the cluster is possible to provide
...
 
-To expose one or more services use the flag `services-configmap`. The format of the data is: `external IP -> namespace/serviceName`.
+To expose one or more services use the flag `services-configmap`. The format of the data is: `external IP -> namespace/serviceName`. Optionally is possible to specify forwarding method using `:` after the service name. The valid options are `NAT` and `DR`. For instance `external IP -> namespace/serviceName:DR`.
+By default if the method is not specified it will use NAT.
+
 This IP must be routable inside the LAN and must be available. 
 By default the IP address of the pods are used to route the traffic. This means that is one pod dies or a new one is created by a scale event the keepalived configuration file will be updated and reloaded.
 
+## [Direct Routing](http://www.linuxvirtualserver.org/VS-DRouting.html)
+
```
Comment (bprashanth) [inline keepalived-vip/README.md]: requirements for DR:
1. hostnetwork
2. L2 connectivity (not possible on clouds. Suggest a small test like arpping MAC on peer machine, telling the user how to check if they have this)
3. nodeselector (I'm not clear on this requirement, can you elaborate? every node that has the selector gets special iptables rules? )

**Comment by aledbf** [comment]
@bprashanth done

File: keepalived-vip/README.md (hunk 2 above)
Comment (bprashanth) [inline keepalived-vip/README.md]: I meant add this to the doc (including the elaboration on nodeselector)

Comment (aledbf) [inline keepalived-vip/README.md]: @bprashanth I removed DR from this PR. Is not stable enough and required additional iptables rules. I will keep it in a branch until I remove the glitches. Sorry about this

Comment (aledbf) [inline keepalived-vip/README.md]: @bprashanth my original goal with this was to fix the source address issues with flannel (the source is not the client IP).
I have another PR after this to use proxy protocol in conjunction with haproxy:
`Internet ->  keepalived VIP -> haproxy with send-proxy`

(this solves the issue with the source ip address)

**Comment by bprashanth** [comment]
LGTM

**Pull request merged**